lib/core/cmap/message_stream.js

```diff
@@ -0,0 +1,177 @@
+const opcodes = require('../wireprotocol/shared').opcodes;
+const compress = require('../wireprotocol/compression').compress;
+const compressorIDs = require('../wireprotocol/compression').compressorIDs;
+const uncompressibleCommands = require('../wireprotocol/compression').uncompressibleCommands;
+const Msg = require('../connection/msg').Msg;
+
+const kDefaultMaxBsonMessageSize = 1024 * 1024 * 16 * 4;
+const kBuffer = Symbol('buffer');
+
+/**
+ * A duplex stream that is capable of reading and writing raw wire protocol messages, with
+ * support for optional compression
+ */
+class MessageStream extends Duplex {
+  constructor(options) {
+    options = options || {};
+    super(options);
+
+    this.bson = options.bson;
+    this.maxBsonMessageSize = options.maxBsonMessageSize || kDefaultMaxBsonMessageSize;
+
+    this[kBuffer] = new BufferList();
+  }
+
+  _write(chunk, _, callback) {
+    this[kBuffer].append(chunk);
+
+    while (this[kBuffer].length >= 4) {
+      const sizeOfMessage = this[kBuffer].readInt32LE(0);
+      if (sizeOfMessage < 0) {
+        callback(new MongoParseError(`Invalid message size: ${sizeOfMessage}`));
+        return;
+      }
+
+      if (sizeOfMessage > this.maxBsonMessageSize) {
+        callback(
+          new MongoParseError(
+            `Invalid message size: ${sizeOfMessage}, max allowed: ${this.maxBsonMessageSize}`
+          )
+        );
+        return;
+      }
+
+      if (sizeOfMessage > this[kBuffer].length) {
+        callback();
+        return;
+      }
+
+      const messageBuffer = this[kBuffer].slice(0, sizeOfMessage);
+      processMessage(this, messageBuffer, callback);
+      this[kBuffer].consume(sizeOfMessage);
+    }
+  }
+
+  _read(/* size */) {
+    // NOTE: This implementation is empty because we explicitly push data to be read
+    //       when `writeMessage` is called.
+    return;
+  }
+
+  writeCommand(command, operationDescription) {
+    // TODO: agreed compressor should live in `StreamDescription`
+    const shouldCompress = operationDescription && !!operationDescription.agreedCompressor;
+    if (!shouldCompress || !canCompress(command)) {
+      this.push(Buffer.concat(command.toBin()));
+      return;
+    }
+
+    // otherwise, compress the message
+    const concatenatedOriginalCommandBuffer = Buffer.concat(command.toBind());
+    const messageToBeCompressed = concatenatedOriginalCommandBuffer.slice(MESSAGE_HEADER_SIZE);
+
+    // Extract information needed for OP_COMPRESSED from the uncompressed message
+    const originalCommandOpCode = concatenatedOriginalCommandBuffer.readInt32LE(12);
+
+    // Compress the message body
+    compress({ options: operationDescription }, messageToBeCompressed, (err, compressedMessage) => {
+      if (err) {
+        operationDescription.cb(err, null);
+        return;
+      }
+
+      // Create the msgHeader of OP_COMPRESSED
+      const msgHeader = Buffer.alloc(MESSAGE_HEADER_SIZE);
```

Comment:
this looks like it was copy-pasta'd so we may not need to do this now, but it seems like we should be able to just do something like:

```js
const msg = Buffer.alloc(MESSAGE_HEADER_SIZE + COMPRESSION_DETAILS_SIZE + compressedMessage.length);
msg.writeInt32LE(
        MESSAGE_HEADER_SIZE + COMPRESSION_DETAILS_SIZE + compressedMessage.length,
        0
      ); // messageLength
msg.writeInt32LE(command.requestId, 4); // requestID
msg.writeInt32LE(0, 8); // responseTo (zero)
msg.writeInt32LE(opcodes.OP_COMPRESSED, 12); // opCode
compressionDetails.writeInt32LE(originalCommandOpCode, 16); // originalOpcode
compressionDetails.writeInt32LE(messageToBeCompressed.length, 20); // Size of the uncompressed compressedMessage, excluding the MsgHeader
compressionDetails.writeUInt8(compressorIDs[operationDescription.agreedCompressor], 24); // compressorID
compressedMessage.copy(msg, 25);
this.push(msg);
```

Remove the need for 3 Buffers + concat. Though tbh I doubt it is that inefficient.

Reply:
I don't have hard numbers for you off-hand, but concatenating multiple buffers in node is actually more efficient than copying between them. In fact, that's part of the impetus for introducing something like `bl`, it attempts to concat as much as possible to provide a "single view" of multiple buffers